kore/src/Kore/Rewrite/Axiom/Matcher.hs

```diff
@@ -81,696 +68,401 @@ import Kore.Internal.Predicate (
+    MatchResult (..),
  )
 import Kore.Simplify.Simplify (
     Simplifier,
  )
 import Kore.Simplify.Simplify qualified as Simplifier
-import Kore.Substitute qualified as Substitute
-import Kore.Variables.Binding
-import Kore.Variables.Fresh qualified as Variables
+import Kore.Substitute
+import Kore.Unparser (
+    unparseToText,
+ )
 import Pair
 import Prelude.Kore
 
 -- * Matching
 
-newtype Constraint variable = Constraint
-    { getConstraint :: Pair (TermLike variable)
-    }
-    deriving stock (Eq)
+{- | Match a TermLike against a pattern as represented by another TermLike.
+ -
+ - Unlike unification, pattern matching is not bidirectional. Variables may
+ - appear in the term being matched, but they will not be unified with the
+ - pattern, and are treated as if they were constants.
+ -
+ - At a high level, the pattern match algorithm maintains two lists. The first
+ - list contains pairs of patterns and terms to match on immediately, and is
+ - initialized with the initial pattern and term. The second list contains
+ - pairs of map or set patterns and terms to be matched on on a deferred basis
+ - after other matching finishes. Matching proceeds by popping a pattern and
+ - a term off the first list, processing it, and performing one of several
+ - actions:
+ -
+ - * failing matching, indicating pattern matching doesn't succeed and
+ -   returning immediately.
+ - * discharging it, indicating that matching locally succeeds and removing it
+ -   from the list
+ - * binding a variable to a term by adding it to the current substitution, and
+ -   removing it from the list
+ - * decomposing it into one or more subpatterns and subterms and adding them
+ -   to the list
+ - * deferring it by moving it to the second list if it is a map or set pattern
+ - * some combination of the above
+ -
+ - Once the first list is empty, we remove one map or set pattern and term from
+ - the second list and perform AC matching. This will perform some combination
+ - of the above actions, potentially adding new elements to the first list.
+ -
+ - The algorithm then repeats this entire process until either matching fails
+ - or both lists are empty, at which time matching succeeds with the
+ - substitution that was accumulated.
+-
+-}
+patternMatch ::
+    SideCondition RewritingVariableName ->
+    TermLike RewritingVariableName ->
+    TermLike RewritingVariableName ->
+    Simplifier (Either Text (MatchResult RewritingVariableName))
+patternMatch sideCondition pat subject =
+    patternMatch' sideCondition [MatchItem pat subject [] Set.empty] MinQueue.empty MultiAnd.top Map.empty
 
-data TermLikeClass
-    = Variables
-    | ConcreteBuiltins
-    | ConstructorAtTop
-    | OtherTermLike
-    | ListBuiltin
-    | AssocCommBuiltin
+data MatchItem = MatchItem (TermLike RewritingVariableName) (TermLike RewritingVariableName) [(SomeVariable RewritingVariableName, SomeVariable RewritingVariableName)] (Set (SomeVariableName RewritingVariableName))
```

Comment:
Can this be made into a record type? I don't know what any of these fields are at a glance.

Reply:
I can, but I gave it a go and it made the code ridiculously more verbose. Are you sure you don't just want me to add a comment explaining what each of the fields are instead?

Reply:
Yes comments would definitely help